### PR TITLE
Fixes #1832 Numeric fields - default values lower than minimum

### LINKED ISSUE #1832: Numeric fields - default values lower than minimum
There is still something annoying about the numeric fields - some of them have their default value lower than the minimum. For example Results per page in Find Users is 0 by default while the minimum is 1, so the browser throws an "error" if the admin doesn't change it.

### PR DESCRIPTION
**Pull request by Stefan-ST**
Fixes #1832

### CODE REVIEW

**Comment by Destroy666x** [comment]
ACP -> Configuration -> Warning System -> Add New Warning Level (and Edit Warning Level)
The default value and minimum of percentage should be 1 to match the description. Same applies to banning/suspension/moderation length - 0 doesn't make any sense as minimum and default value for them.

**Comment by JN-Jones** [comment]
@Stefan-ST

**Comment by JN-Jones** [comment]
@Destroy666x Do you know any other fields? PR looks good to me so far.

**Comment by JN-Jones** [comment]
Add new profilefield:
- maximum length (required field but "0" doesn't make sense. If "0" means ignored it's not protocoled)
- minimum post count
- display order (1 is imho better than 0 here)
- field length

Edit user:
- ICQ number: defaults to 0 if empty
- Date of birth: year defaults to 0 if empty
- Return date: same

Mass mailing:
- Step 3 (define recipients): The three numeric inputs should be empty instead of 0

Add new group promotion:
- Inputs default to 0 instead of empty

**Comment by Destroy666x** [comment]
I already mentioned this in an old issue:

> default to 0 instead of empty

But can we do anything about that besides making the fields text again? That's the magic of numeric fields in browsers - a number is always displayed and there is no way to get rid of it, I think. And since 0s don't do anything in the fields you listed (ICQ is not shown, birthday and return date are hidden too and mass mail conditions with 0 don't work because of weird conditions: https://github.com/mybb/mybb/issues/1766), I'd say there is no need to change them.

**Comment by JN-Jones** [comment]
At least Opera allows empty fields. But I'm fine with leaving them for now, however we'll have the same problem again when someone is going to fix #1766.

**Comment by Destroy666x** [comment]
If it's cross browser (which i recall not to be, but it may have changed), then it's just a matter of editing this line: https://github.com/mybb/mybb/blob/feature/admin/inc/class_form.php#L133 to something like:
```php
if(is_numeric($value))
{
	$value = (float)$value;
}
else
{
	$value = '';
}
$input = "<input type=\"number\" name=\"{$name}\" value=\"{$value}";
```
which will also allow the usage of decimals (no idea why were they skipped).

**Comment by JN-Jones** [comment]
I've tested it with IE, Opera and Chrome and haven't seen any issues so far. Need to retested with your code when I'm home again.

**Comment by Sama34** [comment]
> (no idea why they were rounded on display, it's not an integer field)

Default `min` value is set to `1` in most (if not all) browsers and the value defines other behaviors apart from that of the _minimum_ input value.

https://blog.isotoma.com/2012/03/html5-input-typenumber-and-decimalsfloats-in-chrome/

**Comment by Destroy666x** [comment]
> The `step` attribute value defines those behaviors.

?

I was of course talking about the `(int)` casting.

**Comment by Sama34** [comment]
Duh! Of course.

**Comment by JN-Jones** [comment]
@Stefan-ST 
> Add new profilefield:
- maximum length (required field but "0" doesn't make sense. If "0" means ignored it's not protocoled)
- minimum post count
- display order (1 is imho better than 0 here)
- field length

Can you fix these? The rest should be retested as mentioned above.

**Comment by Stefan-ST** [comment]
@JN-Jones Wouldn't it make more sense to retest the other fields first?

**Comment by JN-Jones** [comment]
> I've tested it with IE, Opera and Chrome and haven't seen any issues so far. Need to retested with your code when I'm home again.

Also tested empty numeric fields with FF and no problems there too. Also checked this change (https://github.com/mybb/mybb/pull/1915#issuecomment-97746000) which seems to work too.

**Comment by Stefan-ST** [comment]
@JN-Jones So we should apply that instead of fixing all the affected fields individually?

**Comment by Destroy666x** [comment]
@Stefan-ST, you can't fix them individually without correcting the function which inserts `0` due to `(int)` casting of empty values.

**Comment by JN-Jones** [comment]
That's up to you - either apply the fix and check all fields or create a separate issue and only fix the non-empty fields here. But remember that 1.8.5 should be released ASAP.

**Comment by Stefan-ST** [comment]
I've updated the PR.

File: admin/inc/class_form.php [hunk 1]
```diff
@@ -130,7 +130,16 @@ function generate_text_box($name, $value="", $options=array())
...
+		if(is_numeric($value))
+		{
+			$value = (float)$value;
+		}
+		else
+		{
+			$value = '';
+		}
+
+		$input = "<input type=\"number\" name=\"{$name}\" value=\"{$value}";
```
Comment (Destroy666x) [inline admin/inc/class_form.php]: Just noticed that a closing " is missing for the `value` attribute

Comment (JN-Jones) [inline admin/inc/class_form.php]: There was a commit which seems to be removed, yes.

**Comment by JN-Jones** [comment]
@Stefan-ST Any comment here? As this already defers 1.8.5 I'd suggest to postpoint this if you don't have time to work on it.

**Comment by JN-Jones** [comment]
@Stefan-ST There were some discussions about wrong default values above, eg the warning points. What's with them?

**Comment by Stefan-ST** [comment]
@JN-Jones I haven't fixed these issues yet.

**Comment by JN-Jones** [comment]
That's why I asked...

**Comment by JN-Jones** [comment]
There's also a "Results per page" field in "Forums -> Attachments" which default to `0`

**Comment by Stefan-ST** [comment]
@JN-Jones "Results per page" is empty for me. From all the fields mentioned here only group promotion fields display a 0 instead of being empty.

**Comment by Destroy666x** [comment]
This should be still fixed:

> ACP -> Configuration -> Warning System -> Add New Warning Level (and Edit Warning Level)
The default value and minimum of percentage should be 1 to match the description. Same applies to banning/suspension/moderation length - 0 doesn't make any sense as minimum and default value for them.

**Comment by Stefan-ST** [comment]
@Destroy666x These fields are not 0 either. They are empty for me which seams to be fine and consistent.

**Comment by Destroy666x** [comment]
But they are required so `1` would be smarter in them by default IMO.

**Comment by Stefan-ST** [comment]
Unless a certain value is recommended we shouldn't set values by default. 1 is actually pretty unlikely for warning points.

Regarding min (and max) values I propose creating a new ticket since it's a much bigger task and low priority.

**Comment by Stefan-ST** [comment]
It would be nice to have this fixed in 1.8.7. ;)

**Comment by Destroy666x** [comment]
> Regarding min (and max) values I propose creating a new ticket since it's a much bigger task and low priority.

Most fields that need to have it already have it: https://github.com/mybb/mybb/pull/1548/files Adding it to these few missing fields isn't big work.

**Comment by Stefan-ST** [comment]
This fixes really annoying issues while min/max isn't important at all. Therefore I prefer to keep this as simple as possible so it doesn't get delayed even further.

**Comment by Cu8eR** [comment]
I agree with @Stefan-ST
